### ISSUE TITLE
Teaching Git interactively in the browser

### ISSUE THREAD

**Issue by QuincyLarson** [freeCodeCamp/freeCodeCamp#16390]
We want to eventually teach Git interactively in the browser. The best example I've found of doing this is this: https://github.com/pcottle/learnGitBranching

It's open source and runs completely client-side.

![learn_git_branching](https://user-images.githubusercontent.com/985197/34470328-ae01c054-eef4-11e7-8e68-52af45391530.png)

This something we could potentially fork and get working within the freeCodeCamp platform, though it would require heavy modification, since it's mostly using a modal-based lessons and a faked Git REPL. 

Anyone interested in spiking on this?

**Comment by vkWeb** [comment]
@QuincyLarson It's great! But we'll need to do a lot of modifications because we'll need to teach our campers from scratch. Like creating a repo, pushing, pulling & committing, etc.

**Comment by saurishkar** [comment]
@QuincyLarson I want to take this up, seems a little challenging to me. Let us also discuss what are the requirements and i believe we need to modify it as per the current format of the challenges on FCC ?
Meanwhile, i will first go through it myself to make myself aware of how to modify it into our own challenges.

**Comment by QuincyLarson** [comment]
@saurishkar Yes - this will indeed be challenging. I have no idea how we'll implement this, but where there's a will, there's a way. We will need to create a "command line" style challenge (right now all our coding challenges use a code editor interface). Then we'll need to get the camper to input the correct command, which will allow them to advance to the next challenge. If you're up for this challenge, let me know.

**Comment by QuincyLarson** [comment]
@bugwheels94 Awesome! Yes - we would welcome your help on this, too! As I told @saurishkar, this will be a big undertaking. But we'll be able to use a command-line interface type challenge for our SQL and Bash challenges, as well. So we'll get a lot of mileage out of your contributions, if this is something you're interested in going forward with.

**Comment by bugwheels94** [comment]
@QuincyLarson Let's begin :)

**Comment by raisedadead** [comment]
@QuincyLarson 
We have had discussed this before, but the one of the best browser tutorials out there is the official

https://try.github.io

**Comment by saurishkar** [comment]
@QuincyLarson As challenging as it seems, it is also interesting how we will get a command line working inside. I thought of mimicking command line using Javascript code, but if i consider the scenario of using it for SQL and Bash challenges in the future, it might be not a good idea.

**Comment by saurishkar** [comment]
@QuincyLarson @raisedadead What I understand is whenever there is a `challengeSeed` key in the seed  .json file, it translates into code for the editor ( i guess `challengeType` has a role to play here ).
So, if i needed to show a terminal instead of an editor, what will i be specifically targeting ?
I researched a little bit and found that we use Code Mirror to hook an editor inside an app.
I found something along these lines : [https://github.com/chjj/tty.js/](url) this might be a good start to look into.

**Comment by raisedadead** [comment]
@saurishkar yup, but preferably a react component console "emulator" is a better choice, like for example which is made on top of the library you have chosen.

**Comment by saurishkar** [comment]
@raisedadead ok , but how will i include it as a seed challenge is what that keeps bugging me?
Any suggestions ?

**Comment by raisedadead** [comment]
Yup. This whole interactive feature will not be trivial to implement without adding supporting infrastructure for the emulator (if we are convinced on one).

We would need a whole stack of client side integration of interactively presenting the code. I'll let Quincy guide more on that though.

**Comment by raisedadead** [comment]
If it were me I would have gone with like we have done for our search a completely separate component that is just dropped in the Nav.

Similarly a completely independent emulator component that can be dropped in replacing the editor would be a way to go.

But that said it would not be trivial IMHO, nonetheless something that is very much possible.
/cc @QuincyLarson maybe we can put this on roadmap.

**Comment by saurishkar** [comment]
@raisedadead ok thanks for the heads-up, i will keep researching and keep you guys posted.

**Comment by QuincyLarson** [comment]
@saurishkar Awesome - thanks for keeping us posted. Adding to our roadmap! We will roll this out once it's ready, but we won't plan for it to be a part of the beta release (we've essentially done a feature freeze until we've shipped beta).

**Comment by bugwheels94** [comment]
Is it only me or everyone faces these many problems at the start? The webpack is not building properly I guess
![Uploading Screenshot 2018-01-12 18:04:51.png…](require is not defined)

**Comment by raisedadead** [comment]
@bugwheels94 can you please create a separate issue.

**Comment by bugwheels94** [comment]
Despite a lot of effort, I cannot run the repo so not sure the "in progress" tag is suited on this given no activity.

**Comment by bugwheels94** [comment]
Created issue https://github.com/freeCodeCamp/freeCodeCamp/issues/16548

**Comment by QuincyLarson** [comment]
@saurishkar We just launched our new learning platform and curriculum. Check it out: learn.freecodecamp.org

Would you still be interested in helping design some interactive in-browser coding challenges?

**Comment by saurishkar** [comment]
@QuincyLarson That's super Awesome !
I will take a look. Thanks

**Comment by QuincyLarson** [comment]
@saurishkar Did you get a chance to look at this? Are you still interested in helping design these in-browser Git coding challenges?

**Comment by saurishkar** [comment]
@QuincyLarson I am afraid haven't got a proper look due to work related timelines. Although i am definitely interested to try my hands on this. I will like to collaborate with others if they have already started on this. Will be a good learning for me.

**Comment by QuincyLarson** [comment]
@saurishkar Thanks for your kind words. Yes - we're just getting started. We want to cover a ton of other web development fundamentals like SQL and Bash as well. But this Git curriculum is a logical place to start. We would welcome your help with it.

**Comment by raisedadead** [comment]
This is open for contributions from anyone.

**Comment by Alex-Cannon** [comment]
@QuincyLarson @raisedadead if this feature is still open I'd like to get started on it.

**Comment by QuincyLarson** [comment]
@Alex-Cannon OK - it's all you. Please keep us posted on your progress.

**Comment by Alex-Cannon** [comment]
@QuincyLarson got it. I'll start on it today. I plan on integrating some sort of visual walk-through, then having them make a git repository. Any suggestions/advice would be appreciated. :) Will keep you posted on my progress.

**Comment by ahmadabdolsaheb** [comment]
Thanks for working on this @Alex-Cannon.
Here is a content list proposal. The following commands are used for contributing to the freecodecamp, it would be great if the git lessons could touch on these materials so campers would feel comfortable contributing to freecodecamp or similar open source projects after finishing the lessons. 

git clone https://github.com/YOUR_USER_NAME/freeCodeCamp.git
git remote add upstream https://github.com/freeCodeCamp/freeCodeCamp.git
git remote -v
git status
git checkout master
git fetch upstream
git reset --hard upstream/master
git push origin master --force
git checkout -b fix/update-guide-for-xyz
git add path/to/my/changed/file.ext
git add .
git commit -m "fix: my short commit message"
git commit --amend
git push origin branch/name-here
git fetch upstream refs/pull/PULLREQ_NUMBER/head:pr-PULLREQ_NUMBER
git checkout pr-PULLREQ_NUMBER
git merge upstream/master
git push git@github.com:USERNAME/REPONAME.git pr-PULLREQ_NUMBER:THEIR_BRANCH_NAME

**Comment by Alex-Cannon** [comment]
@ahmadabdolsaheb nice. I agree that most of these commands should be learned and freeCodeCamp isn't a bad place for people to start learning. Will keep all of you updated.

**Comment by Alex-Cannon** [comment]
Figured out how to add new challenges, new challenge blocks/super blocks, and new challenge types. Below is a pic of the start of the a new git challenge page. 👍 

![](https://github.com/Alex-Cannon/freeCodeCamp/blob/master/update.jpeg?raw=true) 

My plan is to add the editor, display, etc. so users can finish one challenge, using [Learn Git Branching](https://github.com/pcottle/learnGitBranching) as a guide, then add the other challenges. There is so much code in this repo that I plan to pull bits and pieces from it to get it to work in our curriculum, rather than modifying it directly.

Any suggestions on how to best implement [Learn Git Branching](https://github.com/pcottle/learnGitBranching) would be appreciated.

**Comment by QuincyLarson** [comment]
@Alex-Cannon Awesome. Congrats on getting that together so far. I recommend "faking it" at first, and just writing the lessons followed by the commands that people have to type (these could be regex tests). 

This way we get the 80% of the value (teaching people important git commands) with 20% of the work (actually creating an environment where they can stage and unstage files, etc. would be awesome but a lot more work!)

There are some good examples of RegEx-based tests in the AJAX section: https://learn.freecodecamp.org/data-visualization/json-apis-and-ajax/

**Comment by Alex-Cannon** [comment]
As @RandellDawson and @ahmadabdolsaheb suggested, I created two quick wire frames (with the help of @cmccormack). I am in favor of wire frame B. Let me know what you all think. For now, I will move on with the functionality of this feature.

[Wire frame A](https://drive.google.com/open?id=1T3JTy0JqGStBM3ZVNO00sygVqOFxdsxz)
[Wire frame B](https://drive.google.com/open?id=1FUc51ce8ZJoDlzpoHx1AWm-ugr3oZhQ1)

**Comment by Alex-Cannon** [comment]
@QuincyLarson which wire frame should I go with? Or do you have another suggestion?
I also figured out how to create a graph (link below).
[Progress update pic](https://drive.google.com/open?id=1LRuSXnWxN3rOssKGLAKdRCcwePWtb5DD)

**Comment by QuincyLarson** [comment]
@Alex-Cannon This is off to a great start. How are you rendering those diagrams?

Note that we are in the process of redesigning freeCodeCamp's curriculum to use a single-column layout for the challenges, so this may make your life a bit easier. You can simply stack the lesson, diagram, and input terminal, and tests on top of one another.

**Comment by Alex-Cannon** [comment]
@QuincyLarson I'm rendering the diagrams with D3. I am having trouble rendering branch merges though. I haven't worked with Redux or D3, so I'm a little lost, but I'll figure it out. For now I'm going to focus on the terminal + commands.

Here is my suggestion for stacking the Git challenges:
> Instructions
> Git Graph
> Terminal

I'll keep you updated. 🌮

**Comment by tchaffee** [comment]
I don't have time to read this entire thread, but it came up in our call yesterday so I'm dropping this here. This looks very promising. Full git implementation in the browser that works on real git repos. Last commit 24 days ago, 58 issues, 18 PRs. So it's active and supported.

https://isomorphic-git.org/

https://github.com/isomorphic-git/isomorphic-git

**Comment by raisedadead** [comment]
Interesting. They have support for BrowserFS which looks promising. https://isomorphic-git.org/docs/en/fs

**Comment by Alex-Cannon** [comment]
Sorry I have not been active the last month! Moving, looking for a job, etc. I will focus on writing the challenges, and regex tests as @QuincyLarson suggested. Then, we can move on to the visualization.

**Comment by thecodingaviator** [comment]
@QuincyLarson I believe that this might be something we should do for the new curriculum
